packages/nuxt/src/core/nuxt.ts

```diff
@@ -85,16 +86,27 @@ async function initNuxt (nuxt: Nuxt) {
     addWebpackPlugin(UnctxTransformPlugin(nuxt).webpack({ sourcemap: nuxt.options.sourcemap.server || nuxt.options.sourcemap.client }))
   })
 
-  if (!nuxt.options.dev) {
-    const removeFromServer = ['onBeforeMount', 'onMounted', 'onBeforeUpdate', 'onRenderTracked', 'onRenderTriggered', 'onActivated', 'onDeactivated', 'onBeforeUnmount']
-    const removeFromClient = ['onServerPrefetch', 'onRenderTracked', 'onRenderTriggered']
+  // Add composable tree-shaking optimisations
+  const treeShakeOptions: Record<'server' | 'client', TreeShakePluginOptions> = {
+    server: { sourcemap: nuxt.options.sourcemap.server, treeShake: [] },
+    client: { sourcemap: nuxt.options.sourcemap.client, treeShake: [] }
+  }
+  function generateTreeShakeConfig () {
+    treeShakeOptions.server.treeShake = nuxt.options.build.treeShake.server
+    treeShakeOptions.client.treeShake = nuxt.options.build.treeShake.client
+    treeShakeOptions.client = normaliseTreeShakeOptions(treeShakeOptions.client)
+    treeShakeOptions.server = normaliseTreeShakeOptions(treeShakeOptions.server)
+  }
+  nuxt.hook('modules:done', async () => {
+    await generateTreeShakeConfig()
+  })
 
-    // Add tree-shaking optimisations for SSR - build time only
-    addVitePlugin(TreeShakePlugin.vite({ sourcemap: nuxt.options.sourcemap.server, treeShake: removeFromServer }), { client: false })
-    addVitePlugin(TreeShakePlugin.vite({ sourcemap: nuxt.options.sourcemap.client, treeShake: removeFromClient }), { server: false })
-    addWebpackPlugin(TreeShakePlugin.webpack({ sourcemap: nuxt.options.sourcemap.server, treeShake: removeFromServer }), { client: false })
-    addWebpackPlugin(TreeShakePlugin.webpack({ sourcemap: nuxt.options.sourcemap.client, treeShake: removeFromClient }), { server: false })
+  addVitePlugin(TreeShakePlugin.vite(treeShakeOptions.server), { client: false })
+  addVitePlugin(TreeShakePlugin.vite(treeShakeOptions.client), { server: false })
+  addWebpackPlugin(TreeShakePlugin.webpack(treeShakeOptions.server), { client: false })
+  addWebpackPlugin(TreeShakePlugin.webpack(treeShakeOptions.client), { server: false })
```

Comment:
I think we could wait to add these plugins until `modules:done` if that would be helpful. And then the plugin could generate the matcher RegExp when initialised rather than needing a `generateTreeShakeConfig` function.

Reply:
Agreed, that's updated